Add bip-txrelayv2

ariard opened the pull request:
If we find a better way to implement reject unrequested transactions in a backward compatible fashion, I’ll withdraw it.

ML post: https://groups.google.com/g/bitcoindev/c/nWUcXBQbLGU
Bitcoin Core conceptual discussion: https://github.com/bitcoin/bitcoin/pull/30572

murchandamus commented on bip-txrelayv2.mediawiki:
```diff
@@ -0,0 +1,49 @@
+<pre>
+  BIP: XXX
+  Layer: Peer Services
+  Title: Transaction Relay V2
```
Could you perhaps reconsider the title? This sounds somewhat similar to the v2 P2P protocol that was recently deployed. Maybe something along the lines of "Behavior regarding unsolicited Transactions", if I’m getting the gist of what you are thinking about?

ariard replied:
This is intentional to be similar to BIP24 and the upgrade of the V2 P2P protocol. Before the BIP design and implementation, the peer-to-peer protocol was not versioned, neither really encrypted. Similarly, we’re introducing a versioning of the transaction-relay protocol, which a node can use to selectively support policies and mechanisms.

murchandamus commented on bip-txrelayv2.mediawiki:
```diff
@@ -0,0 +1,49 @@
+  Title: Transaction Relay V2
+  Author: Antoine Riard <btc@ariard.me>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0339
```
```suggestion
```

ariard replied:
Corrected with `BIP-0XXX`, in my memory bip339 is the oldest bip related to p2p tx-relay upgrade (as bip338 was never implemented and deployed). So I copy past bip layout from it.

murchandamus commented on bip-txrelayv2.mediawiki:
```diff
@@ -0,0 +1,49 @@
+a new v2 transaction relay protocol. Within this protocol, nodes can negotiate new
+transaction relay mechanisms and policies.
+
+==Specification==
+
+Peers supporting the v2 transaction relay protocol signal support by adverstising
+the 13th bit service flag in the addr p2p messages (`ADDR` and `ADDRV2`).
```
The Specification seems to define how to advertise a new network service, but it should perhaps elaborate what this service entails.

ariard replied:
No, it’s intentional. The mechanisms and policies supported should be described in a distinct BIP, e.g if some class of messages are disabled like it was proposed by BIP338, which is a distinct mechanism than behavior regarding unsolicited transactions.

murchandamus commented on bip-txrelayv2.mediawiki:
```diff
@@ -0,0 +1,49 @@
+Peers supporting the v2 transaction relay protocol signal support by adverstising
+the 13th bit service flag in the addr p2p messages (`ADDR` and `ADDRV2`).
+
+==Backward compatibility==
+
+Older clients remain fully compatible and interoperable after this change.
```
Please be sure to include sufficient detail in the spec or compatibility section to explain why this is the case.

ariard replied:
This is the case - Earliest version behavior can be applied on the linked implementation code is bitcoin core 29.0.

murchandamus reviewed (commented):
Hi @ariard, I was not able to find messages pertaining to this BIP draft on the mailing list. Has this draft been posted to the mailing list?

After a quick read, this document seems to be an early draft. We recommend that BIP authors open an initial pull request against their personal BIPs repository while they are collecting their thoughts and iterating quickly by themselves, and only open a pull request against the main repository after they have shared their proposal with the mailing list and are seeking public review.

ariard commented:
@murchandamus 

> Hi @ariard, I was not able to find messages pertaining to this BIP draft on the mailing list. Has this draft been posted to the mailing list?

An email should have been sent linking this BIP draft to the bitcoin dev mail list.

> After a quick read, this document seems to be an early draft. We recommend that BIP authors open an initial pull request
> against their personal BIPs repository while they are collecting their thoughts and iterating quickly by themselves, and only > open a pull request against the main repository after they have shared their proposal with the mailing list and are seeking
> public review.

My “offensive” self and my “defensive” self after carefully weighing the technical trade-offs have dialogued for long enough and they have come to consensus that this BIP was ready for more review :) I’m seeking public review.

ariard commented:
Updated at [143c993](https://github.com/bitcoin/bips/pull/1663/commits/143c993c5974de14eec92b5a1d317a01056a54e9) with improvements on the BIP.

ariard commented:
Updated at `2a0842b`.

From reading Jon’s comment on the naming issue, and after brainstorming with myself alone, as this BIP can be used to introducing more compartmentalization in the p2p messages (transactions, blocks, addr, etc), it is good to make explicit which version of the p2p protocol, this compartmentalization is building on.

With BIP324 we have a versioning of the protocol, so it can be requested that `NODE_TXRELAY_V2` implies support for `NODE_P2P_V2`.

```
 ==Specification==
 
+Peers supporting the v2 transaction relay protocol SHOULD support the Version 2 P2P Encrypted
+Transport Protocol as specified by BIP324.
+
 Peers supporting the v2 transaction relay protocol signal support by adverstising
 the 13th bit service flag in the addr p2p messages (`ADDR` and `ADDRV2`).
 ```

ariard commented:
Shared an email on the mailing list this day to give an update on this BIP.

Mainly recalling the 3 problems that it wish to improve on:
- transaction-relay as DoS vector
- transaction-propagation deanonymization vector
- transaction-relay throughput overflow attacks on contracting protocols

murchandamus commented:
Hello @ariard, this PR has had unaddressed review comments for over nine months. Are you still working on this?

ariard commented:
> Hello @ariard, this PR has had unaddressed review comments for over nine months. Are you still working on this?

Yes re-implementing my own node in rust(https://bitcoinbackbone.org/). While not there yet on the transaction-relay stuff, I do intend to implement tx-relay v2 in it. So I’ll address the review comments at some point.

murchandamus commented:
Closing this for now, as it has had unaddressed feedback for over 11 months and is not expected to make any progress until an undetermined time in the future. Please feel free to request for this pull request to be reopened or open a new PR when this proposal is closer to ready for publication and more actively developed.

ariard commented on bip-txrelayv2.mediawiki:
```diff
@@ -0,0 +1,51 @@
+processing or alteration in current processing of p2p messages from equivalently upgraded
+peers.
+
+This upgrade mechanism raise issues if new processing of current p2p messages is sanctionned
```
fixed in #1965.

ariard commented on bip-txrelayv2.mediawiki:
```diff
@@ -0,0 +1,51 @@
+Non-upgraded peers and clients can see their transaction broadcast silently failing for a subset
+of theirs opened connections.
+
+We can improve current upgrade mechanism of the transaction relay protocol by introducing a new
```
fixed in #1965
